File: ruby solutions/add_binary.rb
# @param {String} a
# @param {String} b
# @return {String}
def add_binary(a, b)
    res = []
    carry = 0
    i = a.length - 1
    j = b.length - 1

    while i >= 0 or j >= 0 or carry > 0
        if i >= 0
            carry += a[i].to_i
            i -= 1
        end
        if j >= 0
            carry += b[j].to_i
            j -= 1
        end
        res << (carry % 2)
        carry = carry.div(2)
    end
    return res.reverse.join
end

File: python solutions/add_binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        res = []
        carry = 0
        len_a = len(a)-1
        len_b = len(b)-1

        while len_a >= 0 or len_b >= 0 or carry:
            if len_a >= 0:
                carry += int(a[len_a])
                len_a -= 1
            if len_b >= 0:
                carry += int(b[len_b])
                len_b -= 1
            res.append(str(carry % 2))
            carry //= 2
        
        return ''.join(reversed(res))